core/src/main/scala/cats/Apply.scala

```diff
@@ -45,6 +37,9 @@ trait Apply[F[_]] extends Functor[F] with ApplyArityFunctions[F] { self =>
       def F: Apply[F] = self
       def G: Apply[G] = GG
     }
+
+  override def product[A, B](fa: F[A], fb: F[B]): F[(A, B)] = ap(fb)(map(fa)(a => b => (a, b)))
```

Comment:
Would you be willing to reverse this? I.e. leave `product` abstract, and implement `ap` in terms of `product` and `map`? I think the implementation will end up being a bit more efficient.